From bbef40a3d7e53d541ac36ca7bb4b2281396d3b8a Mon Sep 17 00:00:00 2001
From: Ram Dayal <ram.dayal@st.com>
Date: Wed, 13 Nov 2024 09:52:32 +0100
Subject: [PATCH 49/65] phy: stm32-usb2phy: usb2phy is a clock provider of
 48Mhz clock

ck_usb2phyx_48m is generated by usb2phy PLL and used by OTG, USB3DR
and USBH. ck_usb2phyx_48m is available as soon as the PLL is enabled.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 47870911c027bc17659c72dca9f75944f1e594d5 phy: stm32-usb2phy: usb2phy is a clock provider of 48Mhz clock]
Change-Id: I603d7fdf1891ba5dc934ef3710203e9cfaffc5d5
Signed-off-by: Ram Dayal <ram.dayal@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/419008
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 drivers/phy/phy-stm32-usb2phy.c | 154 ++++++++++++++++++++++++++++----
 1 file changed, 138 insertions(+), 16 deletions(-)

diff --git a/drivers/phy/phy-stm32-usb2phy.c b/drivers/phy/phy-stm32-usb2phy.c
index 4fc6c7eefae..06e22bdaf66 100644
--- a/drivers/phy/phy-stm32-usb2phy.c
+++ b/drivers/phy/phy-stm32-usb2phy.c
@@ -11,6 +11,7 @@
 #include <div64.h>
 #include <dm.h>
 #include <dm/device_compat.h>
+#include <dm/device-internal.h>
 #include <dm/lists.h>
 #include <dm/of_access.h>
 #include <fdtdec.h>
@@ -33,6 +34,9 @@
 #define PHY2TRIM1_OFFSET	0x2808
 #define PHY2TRIM2_OFFSET	0x280C
 
+/* USB2PHY_CLK48 */
+#define USB2PHY_CLK48_FREQ	48000000 /* in Hz */
+
 #define PHYCR_REG      1
 
 /* Retention mode enable (active low) */
@@ -79,6 +83,7 @@
 #define SYSCFG_USB2PHYTRIM2_TXPREEMPPULSETUNE_MASK	BIT(2)
 
 struct stm32_usb2phy {
+	struct udevice *dev;
 	struct regmap *regmap;
 	struct clk clk;
 	struct reset_ctl reset;
@@ -216,30 +221,18 @@ static int stm32_usb2phy_regulators_disable(struct phy *phy)
 	return 0;
 }
 
-static int stm32_usb2phy_init(struct phy *phy)
+static int stm32_usb2phy_clk_enable(struct stm32_usb2phy *phy_dev)
 {
 	int ret;
-	struct stm32_usb2phy *phy_dev = dev_get_priv(phy->dev);
-	struct udevice *dev = phy->dev;
+	struct udevice *dev = phy_dev->dev;
+	const struct stm32mp2_usb2phy_hw_data *phy_data = phy_dev->hw_data;
 	unsigned long phyref_rate;
 	u32 phyrefsel;
-	const struct stm32mp2_usb2phy_hw_data *phy_data = phy_dev->hw_data;
-
-	if (phy_dev->init) {
-		phy_dev->init++;
-		return 0;
-	}
-
-	ret = stm32_usb2phy_regulators_enable(phy);
-	if (ret) {
-		dev_err(dev, "can't enable regulators (%d)\n", ret);
-		return ret;
-	}
 
 	ret = clk_enable(&phy_dev->clk);
 	if (ret) {
 		dev_err(dev, "could not enable clock: %d\n", ret);
-		goto error_regl_dis;
+		return ret;
 	}
 
 	phyref_rate = clk_get_rate(&phy_dev->clk);
@@ -261,6 +254,37 @@ static int stm32_usb2phy_init(struct phy *phy)
 		goto error_clk_dis;
 	}
 
+	return 0;
+
+error_clk_dis:
+	clk_disable(&phy_dev->clk);
+
+	return ret;
+}
+
+static int stm32_usb2phy_init(struct phy *phy)
+{
+	int ret;
+	struct stm32_usb2phy *phy_dev = dev_get_priv(phy->dev);
+	struct udevice *dev = phy->dev;
+
+	if (phy_dev->init) {
+		phy_dev->init++;
+		return 0;
+	}
+
+	ret = stm32_usb2phy_regulators_enable(phy);
+	if (ret) {
+		dev_err(dev, "can't enable regulators (%d)\n", ret);
+		return ret;
+	}
+
+	ret = stm32_usb2phy_clk_enable(phy_dev);
+	if (ret) {
+		dev_err(dev, "could not enable clock: %d\n", ret);
+		goto error_regl_dis;
+	}
+
 	ret = reset_deassert(&phy_dev->reset);
 	if (ret) {
 		dev_err(dev, "can't release reset (%d)\n", ret);
@@ -279,6 +303,11 @@ error_regl_dis:
 	return ret;
 }
 
+static int stm32_usb2phy_clk_disable(struct stm32_usb2phy *phy_dev)
+{
+	return clk_disable(&phy_dev->clk);
+}
+
 static int stm32_usb2phy_exit(struct phy *phy)
 {
 	struct stm32_usb2phy *phy_dev = dev_get_priv(phy->dev);
@@ -592,6 +621,8 @@ static int stm32_usb2phy_probe(struct udevice *dev)
 	int ret;
 	u32 phycr;
 
+	phy_dev->dev = dev;
+
 	phy_dev->regmap = syscon_regmap_lookup_by_phandle(dev, "st,syscfg");
 	if (IS_ERR(phy_dev->regmap)) {
 		dev_err(dev, "unable to find regmap\n");
@@ -643,6 +674,16 @@ static int stm32_usb2phy_probe(struct udevice *dev)
 	return 0;
 }
 
+static int stm32_usb2phy_bind(struct udevice *dev)
+{
+	int ret;
+
+	ret = device_bind_driver_to_node(dev, "stm32-usb2phy-clk", "ck_usb2phy_48m",
+					 dev_ofnode(dev), NULL);
+
+	return log_ret(ret);
+}
+
 static const struct udevice_id stm32_usb2phy_of_match[] = {
 	{ .compatible = "st,stm32mp25-usb2phy", .data = (ulong)stm32mp25_usb2phy_hwdata },
 	{ .compatible = "st,stm32mp21-usb2phy", .data = (ulong)stm32mp21_usb2phy_hwdata },
@@ -652,8 +693,89 @@ static const struct udevice_id stm32_usb2phy_of_match[] = {
 U_BOOT_DRIVER(stm32_usb2phy) = {
 	.name = "stm32-usb2phy",
 	.id = UCLASS_PHY,
+	.bind = stm32_usb2phy_bind,
 	.of_match = stm32_usb2phy_of_match,
 	.ops = &stm32_usb2phy_ops,
 	.probe = stm32_usb2phy_probe,
 	.priv_auto	= sizeof(struct stm32_usb2phy),
 };
+
+struct stm32_usb2phy_clk {
+	bool enable;
+	struct clk clkp;
+};
+
+static int stm32_usb2phy_clk48_enable(struct clk *clk)
+{
+	struct stm32_usb2phy_clk *usb2phy_clk = dev_get_priv(clk->dev);
+	struct stm32_usb2phy *usb2phy;
+	int ret;
+
+	if (usb2phy_clk->enable)
+		return 0;
+
+	usb2phy = dev_get_priv(clk->dev->parent);
+
+	/* ck_usb2phy_48m is generated by usb2phy PLL */
+	ret = stm32_usb2phy_clk_enable(usb2phy);
+	if (ret)
+		return ret;
+
+	usb2phy_clk->enable = true;
+
+	return 0;
+}
+
+static int stm32_usb2phy_clk48_disable(struct clk *clk)
+{
+	struct stm32_usb2phy_clk *usb2phy_clk = dev_get_priv(clk->dev);
+	struct stm32_usb2phy *usb2phy;
+	int ret;
+
+	if (!usb2phy_clk->enable)
+		return 0;
+
+	usb2phy = dev_get_priv(clk->dev->parent);
+
+	ret = stm32_usb2phy_clk_disable(usb2phy);
+	if (ret)
+		return ret;
+
+	usb2phy_clk->enable = false;
+
+	return 0;
+}
+
+static unsigned long stm32_usb2phy_clk48_get_rate(struct clk *clk)
+{
+	return USB2PHY_CLK48_FREQ;
+}
+
+static const struct clk_ops stm32_usb2phy_clk48_ops = {
+	.enable = stm32_usb2phy_clk48_enable,
+	.disable = stm32_usb2phy_clk48_disable,
+	.get_rate = stm32_usb2phy_clk48_get_rate,
+};
+
+int usb2phy_clk48_probe(struct udevice *dev)
+{
+	struct stm32_usb2phy_clk *priv = dev_get_priv(dev);
+
+	/* prepare clkp to correctly register clock with CCF */
+	priv->clkp.dev = dev;
+	priv->clkp.id = CLK_ID(dev, 0);
+
+	/* Store back pointer to clk from udevice */
+	/* FIXME: This is not allowed...should be allocated by driver model */
+	dev_set_uclass_priv(dev, &priv->clkp);
+
+	return 0;
+}
+
+U_BOOT_DRIVER(stm32_usb_phyc_clk) = {
+	.name = "stm32-usb2phy-clk",
+	.id = UCLASS_CLK,
+	.ops = &stm32_usb2phy_clk48_ops,
+	.probe = &usb2phy_clk48_probe,
+	.priv_auto = sizeof(struct stm32_usb2phy_clk),
+};
-- 
2.51.0

